src/RF24Log.h

```diff
@@ -0,0 +1,274 @@
+/**
+ * @file RF24Log.h
+ *
+ * Copyright (C)
+ *  2021        Brendan Doherty (2bndy5)
+ *
+ * This General Public License does not permit incorporating your program into
+ * proprietary programs.  If your program is a subroutine library, you may
+ * consider it more useful to permit linking proprietary applications with the
+ * library.  If this is what you want to do, use the GNU Lesser General
+ * Public License instead of this License.
+ */
+
+#ifndef LOGGING_H
+#define LOGGING_H
+
+#include "RF24Log_config.h"
+
+/* macros for testing
+#define ARDUINO // tricking Intellisense
+ */
+
+static const PROGMEM char* const levelDesc[4] = {"DEBUG", "INFO", "WARN", "ERROR"};
+
+/**
+ * @defgroup logLevels Logging Levels
+ * These values are used to delimit certain levels of logging messages
+ * @{
+ */
+/** This level is the default and will output **no** messages of any level */
+#define NOT_SET 0
+/** This level is for debugging outputs and will output messages of @ref DEBUG, @ref INFO, @ref WARN, & @ref ERROR levels */
+#define DEBUG   10
+/** This level is for informative outputs and will output messages of @ref INFO, @ref WARN, & @ref ERROR levels */
+#define INFO    20
+/** This level is for warning outputs and will output messages of @ref WARN & @ref ERROR levels */
+#define WARN    30
+/** This level is for error outputs and will output messages of only @ref ERROR levels */
+#define ERROR   40
+
+/**
+ * @}
+ * A templated class for implementing a single logging object.
+ * Copied @ref logging objects (using RF24Logger(RF24Logger*)) will re-use
+ * the original handler and level.
+ * @see the pre-instantiated @ref logging variable
+ */
+template <typename StreamType>
+class RF24Logger
+{
+private:
+
+    uint8_t level; /** the logging level used to filter logging messages */
+    StreamType* handler; /** the output stream */
+    char _name[40]; /** the logger instance's name (max of 39 characters)*/
+
+public:
+
+    /** Empty constructor. level defaults to @ref NOT_SET, and instance has no handler. */
+    RF24Logger() : level(NOT_SET), handler(nullptr)
```

Comment:
What if someone wants to log to different streams within the same app?

Reply:
If someone develops a library, then he should have no impact to which stream go the log messages. The log messages are just logged "somewhere".

If the end developer wants to write to two different streams, then he needs to create a custom wrapper **handler** where he redirects log messages to those two streams: in Arduino example it could be a **serial port** and **2x16 LCD**.